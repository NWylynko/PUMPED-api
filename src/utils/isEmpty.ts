export const isArrayEmpty = (array: any[]): true | undefined => (array.length === 0 ? true : undefined);

export const isEmpty = (object: any): true | undefined => (!object ? true : undefined);
