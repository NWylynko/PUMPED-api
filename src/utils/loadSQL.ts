import fs from 'fs';
import path from 'path';
import sqlParser from 'sql-template-tag';

const loadSQL = (file: string) => {
  const sql = fs.readFileSync(path.resolve(__dirname, '../../SQL/', `${file}.SQL`), 'utf-8');

  return sqlParser([sql]).sql;
};

export default loadSQL;
