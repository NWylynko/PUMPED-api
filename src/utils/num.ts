const num = (n: string) => parseInt(n, 10);

export default num;
