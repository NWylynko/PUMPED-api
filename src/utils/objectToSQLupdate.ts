import { escape } from 'sqlstring';

const objectToSQLupdate = (fields: any) => {
  let sql = '';

  Object.keys(fields).forEach((partName, index, array) => {
    // if the current item is the length of the array then its the last item
    // the last item cant have a comma
    const comma = index === array.length - 1 ? '' : ',';

    sql += `${escape(partName)} = ${escape(fields[partName])}${comma} `;
  });

  return sql;
};

export default objectToSQLupdate;
