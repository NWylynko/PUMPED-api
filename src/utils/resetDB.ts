import { promises as fs } from 'fs';
import db, { SQL } from '../db';

export const createTables = () => Promise.all([
  db.run(SQL.tables.Brand),
  db.run(SQL.tables.Collection),
  db.run(SQL.tables.Colour),
  db.run(SQL.tables.ColourImage),
  db.run(SQL.tables.Customer),
  db.run(SQL.tables.Image),
  db.run(SQL.tables.Order),
  db.run(SQL.tables.OrderItem),
  db.run(SQL.tables.Review),
  db.run(SQL.tables.Section),
  db.run(SQL.tables.Shoe),
  db.run(SQL.tables.ShoeTag),
  db.run(SQL.tables.Stock),
  db.run(SQL.tables.Style),
  db.run(SQL.tables.Tag),
  db.run(SQL.tables.WishList),
]);

const resetDB = async () => {
  await db.close();
  await fs.unlink(`./main.${process.env.NODE_ENV}.db`);
  await db.open();
  await createTables();
};

export const addTestData = () => db.exec(SQL.testData);

export const resetDBForTest = async () => {
  await db.close();
  await db.open();
  await createTables();
  await db.exec(SQL.testData);
};

export default resetDB;
