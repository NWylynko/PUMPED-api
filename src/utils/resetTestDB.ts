import { promises as fs } from 'fs';
import db from '../db';

const reset = async () => {
  await db.close();
  await fs.unlink('./main.test.db');
  await fs.copyFile('./main.testData.db', './main.test.db');
  await db.open();
};

export default reset;
