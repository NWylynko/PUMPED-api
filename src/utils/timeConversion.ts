function timeConversion(milliseconds: number) {
  // adapted from https://stackoverflow.com/a/32180863

  const seconds = (milliseconds / 1000);
  const minutes = (milliseconds / (1000 * 60));
  const hours = (milliseconds / (1000 * 60 * 60));
  const days = (milliseconds / (1000 * 60 * 60 * 24));

  if (seconds < 60) {
    return `${Math.floor(seconds)} Seconds`;
  } if (minutes < 60) {
    return `${Math.floor(minutes)} Minutes`;
  } if (hours < 24) {
    return `${Math.floor(hours)} Hours`;
  }
  return `${Math.floor(days)} Days`;
}

export default timeConversion;
