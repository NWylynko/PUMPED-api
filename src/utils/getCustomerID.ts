import { Request } from 'express';
import num from './num';

const getCustomerID = (req: Request): number => {
  const CustomerID = req.header('CustomerID');

  if (!CustomerID) {
    throw new Error('no customer id supplied');
  }

  return num(CustomerID);
};

export default getCustomerID;
