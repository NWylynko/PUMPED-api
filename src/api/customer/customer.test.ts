// eslint-disable-next-line import/no-extraneous-dependencies
import supertest from 'supertest';
import express from '../../app';
import { resetDBForTest } from '../../utils/resetDB';

const app = supertest(express);

beforeEach(async (done) => {
  await resetDBForTest();
  done();
});

describe('GET Customer', () => {
  [1, 2, 3, 4].forEach(async (CustomerID) => {
    it(`gets Customer for ID (${CustomerID})`, async () => {
      const response = await app.get(encodeURI(`/customer/${CustomerID}`));

      expect(response.status).toMatchSnapshot();

      const json = JSON.parse(response.text);

      expect(json).toMatchSnapshot();
    });
  });
});

describe('POST add a new Customer', () => {
  it('no data', async () => {
    const customer = {};

    const addCustomer = await app
      .post('/customer/')
      .type('json')
      .send(JSON.stringify(customer));

    expect(addCustomer.status).toMatchSnapshot();

    const addCustomerJson = JSON.parse(addCustomer.text);

    expect(addCustomerJson).toMatchSnapshot();

    const getCustomer = await app.get('/customer/4');

    expect(getCustomer.status).toMatchSnapshot();

    const getCustomerJson = JSON.parse(getCustomer.text);

    expect(getCustomerJson).toMatchSnapshot();
  });

  it('not enough data', async () => {
    const customer = {
      firstName: 'jacky',
    };

    const addCustomer = await app
      .post('/customer/')
      .type('json')
      .send(JSON.stringify(customer));

    expect(addCustomer.status).toMatchSnapshot();

    const addCustomerJson = JSON.parse(addCustomer.text);

    expect(addCustomerJson).toMatchSnapshot();

    const getCustomer = await app.get('/customer/4');

    expect(getCustomer.status).toMatchSnapshot();

    const getCustomerJson = JSON.parse(getCustomer.text);

    expect(getCustomerJson).toMatchSnapshot();
  });

  it('correct amount of data', async () => {
    const customer = {
      firstName: 'big',
      lastName: 'lad',
    };

    const addCustomer = await app
      .post('/customer/')
      .type('json')
      .send(JSON.stringify(customer));

    expect(addCustomer.status).toMatchSnapshot();

    const addCustomerJson = JSON.parse(addCustomer.text);

    expect(addCustomerJson).toMatchSnapshot();

    const getCustomer = await app.get('/customer/4');

    expect(getCustomer.status).toMatchSnapshot();

    const getCustomerJson = JSON.parse(getCustomer.text);

    expect(getCustomerJson).toMatchSnapshot();
  });
});

describe('PATCH update a Customer', () => {
  const customers = [1, 4]; // registers customer and non existent customer

  const customerData = [
    { name: 'no data', data: {} },
    {
      name: 'only one field',
      data: {
        firstName: 'jacky',
      },
    },
    {
      name: 'all fields',
      data: {
        firstName: 'big',
        lastName: 'lad',
      },
    },
  ];

  customers.forEach((customerID) => {
    customerData.forEach(({ name, data }) => {
      it(`${customerID === 1 ? '' : 'no '}customer, ${name}`, async () => {
        const updateCustomer = await app
          .patch(`/customer/${customerID}`)
          .type('json')
          .send(JSON.stringify(data));

        expect(updateCustomer.status).toMatchSnapshot();

        const updateCustomerJson = JSON.parse(updateCustomer.text);

        expect(updateCustomerJson).toMatchSnapshot();

        const getCustomer = await app.get(`/customer/${customerID}`);

        expect(getCustomer.status).toMatchSnapshot();

        const getCustomerJson = JSON.parse(getCustomer.text);

        expect(getCustomerJson).toMatchSnapshot();
      });
    });
  });
});

describe('DELETE remove a Customer', () => {
  [1, 2, 3, 4].forEach(async (CustomerID) => {
    it(`removes Customer for ID (${CustomerID})`, async () => {
      const response = await app.delete(encodeURI(`/customer/${CustomerID}`));

      expect(response.status).toMatchSnapshot();

      const json = JSON.parse(response.text);

      expect(json).toMatchSnapshot();

      const getCustomer = await app.get(`/customer/${CustomerID}`);

      expect(getCustomer.status).toMatchSnapshot();

      const getCustomerJson = JSON.parse(getCustomer.text);

      expect(getCustomerJson).toMatchSnapshot();
    });
  });
});
