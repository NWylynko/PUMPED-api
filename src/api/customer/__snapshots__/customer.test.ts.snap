// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE remove a Customer removes Customer for ID (1) 1`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (1) 2`] = `
Object {
  "data": Object {
    "CustomerID": "1",
  },
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (1) 3`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (1) 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (2) 1`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (2) 2`] = `
Object {
  "data": Object {
    "CustomerID": "2",
  },
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (2) 3`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (2) 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (3) 1`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (3) 2`] = `
Object {
  "data": Object {
    "CustomerID": "3",
  },
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (3) 3`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (3) 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (4) 1`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (4) 2`] = `
Object {
  "data": Object {
    "CustomerID": "4",
  },
  "success": true,
}
`;

exports[`DELETE remove a Customer removes Customer for ID (4) 3`] = `200`;

exports[`DELETE remove a Customer removes Customer for ID (4) 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`GET Customer gets Customer for ID (1) 1`] = `200`;

exports[`GET Customer gets Customer for ID (1) 2`] = `
Object {
  "data": Object {
    "firstName": "bobby",
    "lastName": "boo",
  },
  "success": true,
}
`;

exports[`GET Customer gets Customer for ID (2) 1`] = `200`;

exports[`GET Customer gets Customer for ID (2) 2`] = `
Object {
  "data": Object {
    "firstName": "pippy",
    "lastName": "nippy",
  },
  "success": true,
}
`;

exports[`GET Customer gets Customer for ID (3) 1`] = `200`;

exports[`GET Customer gets Customer for ID (3) 2`] = `
Object {
  "data": Object {
    "firstName": "whip",
    "lastName": "nah",
  },
  "success": true,
}
`;

exports[`GET Customer gets Customer for ID (4) 1`] = `200`;

exports[`GET Customer gets Customer for ID (4) 2`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`PATCH update a Customer customer, all fields 1`] = `200`;

exports[`PATCH update a Customer customer, all fields 2`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "firstName": "big",
    "lastName": "lad",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer customer, all fields 3`] = `200`;

exports[`PATCH update a Customer customer, all fields 4`] = `
Object {
  "data": Object {
    "firstName": "big",
    "lastName": "lad",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer customer, no data 1`] = `500`;

exports[`PATCH update a Customer customer, no data 2`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`PATCH update a Customer customer, no data 3`] = `200`;

exports[`PATCH update a Customer customer, no data 4`] = `
Object {
  "data": Object {
    "firstName": "bobby",
    "lastName": "boo",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer customer, only one field 1`] = `200`;

exports[`PATCH update a Customer customer, only one field 2`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "firstName": "jacky",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer customer, only one field 3`] = `200`;

exports[`PATCH update a Customer customer, only one field 4`] = `
Object {
  "data": Object {
    "firstName": "jacky",
    "lastName": "boo",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer no customer, all fields 1`] = `200`;

exports[`PATCH update a Customer no customer, all fields 2`] = `
Object {
  "data": Object {
    "CustomerID": "4",
    "firstName": "big",
    "lastName": "lad",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer no customer, all fields 3`] = `200`;

exports[`PATCH update a Customer no customer, all fields 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`PATCH update a Customer no customer, no data 1`] = `500`;

exports[`PATCH update a Customer no customer, no data 2`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`PATCH update a Customer no customer, no data 3`] = `200`;

exports[`PATCH update a Customer no customer, no data 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`PATCH update a Customer no customer, only one field 1`] = `200`;

exports[`PATCH update a Customer no customer, only one field 2`] = `
Object {
  "data": Object {
    "CustomerID": "4",
    "firstName": "jacky",
  },
  "success": true,
}
`;

exports[`PATCH update a Customer no customer, only one field 3`] = `200`;

exports[`PATCH update a Customer no customer, only one field 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`POST add a new Customer correct amount of data 1`] = `200`;

exports[`POST add a new Customer correct amount of data 2`] = `
Object {
  "data": Object {
    "firstName": "big",
    "lastName": "lad",
  },
  "success": true,
}
`;

exports[`POST add a new Customer correct amount of data 3`] = `200`;

exports[`POST add a new Customer correct amount of data 4`] = `
Object {
  "data": Object {
    "firstName": "big",
    "lastName": "lad",
  },
  "success": true,
}
`;

exports[`POST add a new Customer no data 1`] = `500`;

exports[`POST add a new Customer no data 2`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`POST add a new Customer no data 3`] = `200`;

exports[`POST add a new Customer no data 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`POST add a new Customer not enough data 1`] = `500`;

exports[`POST add a new Customer not enough data 2`] = `
Object {
  "error": true,
  "message": "SQLITE_CONSTRAINT: NOT NULL constraint failed: Customer.lastName",
  "params": Array [
    "jacky",
    null,
  ],
  "result": Object {
    "code": "SQLITE_CONSTRAINT",
    "errno": 19,
  },
  "sql": " INSERT INTO Customer (\\"firstName\\",\\"lastName\\" ) VALUES (?,? ) ",
}
`;

exports[`POST add a new Customer not enough data 3`] = `200`;

exports[`POST add a new Customer not enough data 4`] = `
Object {
  "empty": true,
  "success": true,
}
`;
